File: PythonCodes/Fig09_10.py
#%%
"""
Created on Jan 22 2019
Monte Carlo paths for the Jacobi correlation model
@author: Lech A. Grzelak
"""
import numpy as np
import matplotlib.pyplot as plt

def GeneratePathsJacobi(NoOfPaths,NoOfSteps,T,rho0,kappa,gamma,mu):    
    Z = np.random.normal(0.0,1.0,[NoOfPaths,NoOfSteps])
    W = np.zeros([NoOfPaths, NoOfSteps+1])
    Rho = np.zeros([NoOfPaths, NoOfSteps+1])
    Rho[:,0]=rho0
    time = np.zeros([NoOfSteps+1])
        
    dt = T / float(NoOfSteps)
    for i in range(0,NoOfSteps):

        # Making sure that samples from a normal have mean 0 and variance 1

        if NoOfPaths > 1:
            Z[:,i] = (Z[:,i] - np.mean(Z[:,i])) / np.std(Z[:,i])
        W[:,i+1] = W[:,i] + np.power(dt, 0.5)*Z[:,i]
        Rho[:,i+1] = Rho[:,i] + kappa*(mu - Rho[:,i]) * dt + gamma* np.sqrt(1.0 -Rho[:,i]*Rho[:,i]) * (W[:,i+1]-W[:,i])
        
        # Handling of the boundary conditions to ensure paths stay within the [-1,1] range

        Rho[:,i+1] = np.maximum(Rho[:,i+1],-1.0)
        Rho[:,i+1] = np.minimum(Rho[:,i+1],1.0)
        time[i+1] = time[i] +dt
        
    # Output

    paths = {"time":time,"Rho":Rho}
    return paths

def mainCalculation():
    NoOfPaths = 15
    NoOfSteps = 500
    T         = 5
    kappa     = 0.25
    rho0      = 0.0
    kappa     = 0.1
    gamma     = 0.6
    mu        = 0.5
    
    # Check the boundaries:

    if kappa > np.max([gamma*gamma/(1.0-mu),gamma*gamma/(1.0+mu)]):
        print("Boundry is NOT attainable")
    else:
        print("Boundry is attainable")
        
    Paths = GeneratePathsJacobi(NoOfPaths,NoOfSteps,T,rho0,kappa,gamma,mu)   
    timeGrid = Paths["time"]
    Rho = Paths["Rho"]
    
    plt.figure(1)
    plt.plot(timeGrid, np.transpose(Rho),'b')   
    plt.grid()
    plt.xlabel("time")
    plt.ylabel("Rho(t)")
           
mainCalculation()

File: MatlabCodes/Chapter 09/Fig09_10.m
function Jacobi_Paths
clc;close all;
NoOfPaths  =5000;
NoOfSteps  = 10; 
T          = 5;
rho0      = 0.0;
kappa     = 0.1;
gamma     = 0.6;
mu        = 0.5;

figure(1)
randn('seed',1)
[Rho,timeGrid] = GeneratePathsJackobi(NoOfPaths,NoOfSteps,T,rho0,mu,kappa,gamma);
plot(timeGrid,Rho,'color',[0,0.45,0.75],'linewidth',1.5)
axis([0,T,min(min(Rho)),max(max(Rho))])
grid on
xlabel('time')
ylabel('Rho(t)')

figure(2)
h1 = histogram(Rho(:,end),50);
h1.FaceColor=[1,1,1];
grid on
xlabel('Error at time T')
title('Rho')
grid on;
function [Rho,time] = GeneratePathsJackobi(NoOfPaths,NoOfSteps,T,rho0,mu,kappa,gamma)
Rho=zeros(NoOfPaths,NoOfSteps);
Rho(:,1) = rho0;

% Random noise

Z=random('normal',0,1,[NoOfPaths,NoOfSteps]);
W=zeros([NoOfPaths,NoOfSteps]);

dt = T / NoOfSteps;
time = zeros([NoOfSteps+1,1]);
for i=1:NoOfSteps
    if NoOfPaths>1
        Z(:,i)   = (Z(:,i) - mean(Z(:,i))) / std(Z(:,i));
    end
    W(:,i+1)  = W(:,i) + sqrt(dt).*Z(:,i);
    Rho(:,i+1) = Rho(:,i) + kappa*(mu - Rho(:,i)) * dt + gamma* sqrt(1.0 -Rho(:,i).*Rho(:,i)) .* (W(:,i+1)-W(:,i));
        
    % Handling of the boundary conditions to ensure that paths stay in the (-1,1] range

    Rho(:,i+1) = max(Rho(:,i+1),-1.0);
    Rho(:,i+1) = min(Rho(:,i+1),1.0);
    time(i+1) = time(i) + dt;
end
